SI-6412 some fixes for reflection leaks

xeno-by opened the pull request.

xeno-by committed:
SI-6412 fixes leaks in Types.uniques

This is the most blatant leak in reflection. There are others, but their impact
is much smaller, therefore we'll fix them later, after 2.10.0-final.

For more information, see https://issues.scala-lang.org/browse/SI-6412 and
http://groups.google.com/group/scala-internals/browse_thread/thread/eabcf3d406dab8b2

xeno-by commented:
review @paulp

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/575/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1283/

paulp commented on src/partest/scala/tools/partest/MemoryTest.scala:
```diff
@@ -0,0 +1,37 @@
+package scala.tools.partest
+
+abstract class MemoryTest extends App {
```
For a test like this, stay far away from sketchy features like DelayedInit.  It saves you one line and cannot possibly reduce the number of confounding variables.

paulp commented on src/reflect/scala/reflect/runtime/SynchronizedTypes.scala:
```diff
@@ -11,7 +14,12 @@ trait SynchronizedTypes extends internal.Types { self: SymbolTable =>
-  override def unique[T <: Type](tp: T): T = uniqueLock.synchronized { super.unique(tp) }
+  private val uniques = WeakHashMap[Type, WeakReference[Type]]()
+  override def unique[T <: Type](tp: T): T = uniqueLock.synchronized {
+    var result = uniques.getOrElseUpdate(tp, new WeakReference(tp)).get.asInstanceOf[T]
```
Unfortunately getOrElseUpdate is very slow compared to writing the same logic by hand.

paulp commented on src/reflect/scala/reflect/runtime/SynchronizedTypes.scala:
```diff
@@ -11,7 +14,12 @@ trait SynchronizedTypes extends internal.Types { self: SymbolTable =>
+    if (result == null) { uniques(tp) = new WeakReference(tp); result = tp }
+    result
+  }
 
```
The logic for this should be more like

```
val result = if (uniques contains tp) uniques(tp).get else null
if (result ne null) result
else {
  uniques(tp) = new WeakReference(tp)
  tp
}
```

No vars, no getOrElseUpdate.

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1283/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/575/

xeno-by commented:
PLS REBUILD ALL

xeno-by commented:
This pull request depends on https://github.com/scala/scala/pull/1378, because it touches `Types.uniques`, which got a fix in #1378.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/576/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1284/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/576/

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1284/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/578/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/578/

xeno-by committed:
SI-6412 alleviates leaks in toolboxes

Turns importer caches into fully weak hash maps, and also applies
manual cleanup to toolboxes every time they are used.

It's not enough, because reflection-mem-typecheck test is still leaking
at a rate of ~100kb per typecheck, but it's much better than it was before.
We'll fix the rest later, after 2.10.0-final.

For more information, see https://issues.scala-lang.org/browse/SI-6412 and
http://groups.google.com/group/scala-internals/browse_thread/thread/eabcf3d406dab8b2

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/579/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1286/

jsuereth commented on src/partest/scala/tools/partest/MemoryTest.scala:
```diff
@@ -0,0 +1,38 @@
+    1 to 5 foreach (_ => stressTestIter())
+    val reference = memUsage()
+    1 to 5 foreach (_ => stressTestIter())
+    1 to 5 foreach (_ => rt.gc())
```
There's no guarantee this will ever actual call the gc.  One thing that used to work well was lowering the priority of every runtime thread and bump the priority of the GC.  With concurrent mark+sweep, I'm not sure that really helps.

While we don't want this test to be flaky, I think this is ok for now.   I think longer term maybe we can use re-use a bytecode-rewriter to tag/track memory allocations (if one of the profiles has its agents OSS)

xeno-by committed:
fixes mirrorThatLoaded in JavaMirrors

`mirrorThatLoaded(sym: Symbol): Mirror` in JavaMirrors used to iterate over
`lazy val mirrors = new WeakHashMap[ClassLoader, WeakReference[JavaMirror]]()`
to find out what mirror has loaded a certain symbol.

It worked okay until yesterday when I noticed failing tests, which crashed
when weak references in mirrors were dereferenced with get.get. Apparently
some mirrors were collected, and the logic in JavaMirror didn't account for
that possibility.

When fixing this bug, I noticed that mirrors can become unreachable even
if there are still reachable symbols created by those mirrors. That doesn't
make sense, therefore I fixed this bug as well by introducing a strong ref
from root symbols to the enclosing mirror. Therefore, any active symbol
will have a strong reference to the enclosing mirror by the virtue of the
owner chain.

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1286/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/579/

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/580/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1287/

xeno-by commented:
if noone has objections, let's merge?

scala-jenkins commented:
Fine with me ....

paulp commented:
Looks like progress in any case.

gkossakowski commented:
It looks like this pull request breaks rangepos tests. I'm checking it now locally. //cc @hubertp

hubertp commented:
I think some confusion crept it. This PR doesn't break rangepos. It breaks normal tests.

xeno-by commented:
Fixed in #1391

xeno-by commented:
@hubertp @gkossakowski @odersky Funny thing is that the tests never passed to begin with. Though the kitteh never announced the failure.

So when I suggested to merge, I was under the impression that the tests pass. You know, like an optical illusion. 

I am sorry guys to have wasted your time.
